### ISSUE TITLE
false positive in @NonNegative check

### ISSUE THREAD

**Issue by zyadahmed** [typetools/checker-framework#7008]
```
 final int referenceIndex = extractReferencePart(content);
        Optional<JavadocTag> validTag = Optional.empty();

        if (referenceIndex != -1) {
            final String referenceString;
            if (referenceIndex == 0) {
                referenceString = content;
            }
            else {
                referenceString = content.substring(0, referenceIndex);
            }
```
The Checker Framework incorrectly reports an error for `content.substring(0, referenceIndex)`, stating that `referenceIndex` must be a @NonNegative int. However, the condition `if (referenceIndex != -1)` ensures that only possible negative return value is handled before execution.
Is there a way for Checker Framework to handle this inference internally and recognize the safety of referenceIndex?

```
 <message>incompatible argument for parameter endIndex of String.substring.</message>
    <lineContent>referenceString = content.substring(0, referenceIndex);</lineContent>
    <details>
      found   : int
      required: @NonNegative int
   </details>
```

**Comment by mernst** [comment]
This is not a false positive.  If you do not specify `extractReferencePart()` otherwise, then the Checker Framework conservatively assumes that `extractReferencePart()` might return any value, including -2.

If you specify `extractReferencePart()` to return a value that is -1 or larger, then type-checking passes:

```java
import org.checkerframework.checker.index.qual.IndexOrLow;

class Issue7008 {

  @IndexOrLow("#1") int extractReferencePart(String s) {
    // ... The body doesn't matter when type-checking `m`; only the specification matters.
    return s.indexOf("reference");
  }

  void m(String content) {
    final int referenceIndex = extractReferencePart(content);

    if (referenceIndex != -1) {
      final String referenceString;
      if (referenceIndex == 0) {
        referenceString = content;
      } else {
        referenceString = content.substring(0, referenceIndex);
      }
    }
  }
}
```
